controllers/k8ssandra/datacenters.go

```diff
@@ -37,7 +39,7 @@ func (r *K8ssandraClusterReconciler) reconcileDatacenters(ctx context.Context, k
 	}
 
 	// Reconcile CassandraDatacenter objects only
-	for _, dcTemplate := range kc.Spec.Cassandra.Datacenters {
+	for idx, dcTemplate := range kc.Spec.Cassandra.Datacenters {
```

Comment:
Regarding the handling of system replication:

On line 27 we define or retrieve the initial system replication:

```go
systemReplication, err := r.checkSystemReplication(ctx, kc, logger)
```

Then on line 57 we apply that replication to each `dcConfig`:

```go
cassandra.ApplySystemReplication(dcConfig, *systemReplication)
```

This is OK if the dc being reconciled is an initial one. But if it is a dc that is being added after the cluster is initialized, then this means that we are applying a wrong system replication to its config, and a wrong one that does not include the dc being created. Shouldn't we fix this right here?

Instead, it seems this is going to be "fixed" later on on line 182:

```go
if recResult := r.updateReplicationOfSystemKeyspaces(ctx, kc, mgmtApi, logger); 
```

Indeed the `updateReplicationOfSystemKeyspaces` method behaves like a no-op for initial dcs, but triggers a change in the system keyspaces replication when adding a new dc, which eventually fixes the wrong initial system replication that the dc was created with.

2 questions here:

* Wouldn't it be simpler to just create the new dc with the right replication and let management-api do its magic? 
* If that is not possible and we need to keep things as they are: calling `updateReplicationOfSystemKeyspaces` on an initial dc seems like a waste of time, since we know the replication is OK and so `EnsureKeyspaceReplication` will do nothing. Shouldn't we only call it for dcs being added to an initialized cluster?

Reply:
Calling `cassandra.ApplySystemReplication(dcConfig, *systemReplication)` at line 57 for new DCs being added to an existing cluster does not result in the wrong system replication. Let's say we have a cluster with a single DC, dc1. System replication will be configured for it with these JVM system properties:

* `cassandra.system_distributed_replication_dc_names=dc1`
* `cassandra.system_distributed_replication_per_dc=3

Now let's say we after the cluster is fully initialized we add dc2. It will get those same system properties, which means we aren't replicating to dc2 yet. That's ok. As you point out, we could configure the system properties for dc2 so we have `cassandra.system_distributed_replication_dc_names=dc1,dc2`. I don't like this because now instead of managing one initial system replication for the whole cluster, we would have to manage it per DC.

We still have to call `updateReplicationOfSystemKeyspaces` on an initial DC as the replication could change. If we start with a single node, we will have RF = 1. If we later scale dc1 to 3 nodes, then system keyspaces will get updated to RF = 3 only when `updateReplicationOfSystemKeyspaces` is called.

This is confusing and messy. I think it will be a lot better if we do what I describe in https://github.com/k8ssandra/k8ssandra-operator/pull/262#discussion_r789200931.

Reply:
> It will get those same system properties, which means we aren't replicating to dc2 yet. That's ok. 

👍 

> We still have to call updateReplicationOfSystemKeyspaces on an initial DC as the replication could change.

👍